### ISSUE TITLE
Fix dataset_to_point_list in util.py

### ISSUE THREAD

**Issue by asb219** [pymc-devs/pymc#5089]
Dear developers,

I would like to report an issue with the function `dataset_to_point_list` in `pymc/util.py`. I am proposing some small changes to this function which I believe would fix the bug. Do you approve of these changes? Or perhaps do you have better ideas for how to fix the issue?

Hopefully this can be sorted out quickly.

Thanks in advance,
Alex

## Bug description

The argument for `pm.sample_posterior_predictive` can not be a trace in form of an arviz InferenceData which was burned or thinned.

How to (hopefully) reproduce the error:
```python
with any_pymc3_model:
    trace = pm.sample(draws=1000, return_inferencedata=True)
    trace = trace.sel(draw=slice(200,None)) # burn the first 200 draws
    predict = pm.sample_posterior_predictive(trace) # will raise an IndexError
```
I get `IndexError: index 800 is out of bounds for axis 1 with size 800`. This error is raised while running the function `dataset_to_point_list` in `pymc3/util.py`.

<details><summary>Complete error traceback</summary>

```python
Traceback (most recent call last):
    predict = pm.sample_posterior_predictive(trace)
  File "/home/asb219/storage/miniconda3/envs/cctest/lib/python3.9/site-packages/pymc3/sampling.py", line 1653, in sample_posterior_predictive
    _trace = dataset_to_point_list(trace.posterior)
  File "/home/asb219/storage/miniconda3/envs/cctest/lib/python3.9/site-packages/pymc3/util.py", line 287, in dataset_to_point_list
    points.append({vn: s[c, d] for vn, s in _samples.items()})
  File "/home/asb219/storage/miniconda3/envs/cctest/lib/python3.9/site-packages/pymc3/util.py", line 287, in <dictcomp>
    points.append({vn: s[c, d] for vn, s in _samples.items()})
IndexError: index 800 is out of bounds for axis 1 with size 800
```

</details>

<details><summary>Quick fix</summary>

Avoid the error by using MultiTrace instead of InferenceData:

```python
with any_pymc3_model:
    trace = pm.sample(draws=1000, return_inferencedata=False)
    trace = trace[200:] # burn the first 200 draws
    prediction = pm.sample_posterior_predictive(trace) # no problem
```

</details>

 ## Proposed fix

I would like to propose a "correction" to the function `dataset_to_point_list` in `pymc/util.py` which would fix the problem described above.

Here is what the function currently looks like on Github:
```python
def dataset_to_point_list(ds: xarray.Dataset) -> List[Dict[str, np.ndarray]]:
    # grab posterior samples for each variable
    _samples: Dict[str, np.ndarray] = {vn: ds[vn].values for vn in ds.keys()}
    # make dicts
    points: List[Dict[str, np.ndarray]] = []
    vn: str
    s: np.ndarray
    for c in ds.chain:
        for d in ds.draw:
            points.append({vn: s[c, d] for vn, s in _samples.items()})
    # use the list of points
    return points
```
In the example given in the "Bug description" section above, **we get the `IndexError` because `ds.draw` takes values from 200 to 1000, and not 0 to 800.**

Proposed changes:
```python
def dataset_to_point_list(ds: xarray.Dataset) -> List[Dict[str, np.ndarray]]:
    ...
    # Either this:
    for c in ds.chain:
        for d in ds.draw:
            points.append({vn: s.sel(chain=c, draw=d) for vn, s in _samples.items()})

    # Or that:
    for c in range(ds.chain.size):
        for d in range(ds.draw.size):
            points.append({vn: s[c, d] for vn, s in _samples.items()})
    ...
```

Though I'm sure you developers can probably think of a better solution.

## Versions and main components

* PyMC/PyMC3 Version: 3.11.4 
* Python Version: 3.9 (also tested with 3.8)
* How did you install PyMC/PyMC3: conda

**Comment by 5hv5hvnk** [comment]
I think the issue is resolved in PyMC version `v4.1.1`

**Comment by OriolAbril** [comment]
As @5hv5hvnk pointed, this has been fixed already in the latest release, you can see the updated code in https://github.com/pymc-devs/pymc/blob/main/pymc/util.py#L234-L243. The issue was indeed that label based and position based indexing were mixed, now only label based indexing is used.

I will therefore close the issue, you will need to update to 4.x to be able to use burned  or thinned inferencedata objects as input.
